### PR TITLE
[WIP] fix(fuzz): deduplicate fuzz inputs

### LINKED ISSUE #3521: feat: deduplicate fuzz inputs
### Component

Forge

### Describe the feature you would like

Currently foundry's fuzz inputs are quite wasteful. Here's a [spreadsheet](https://docs.google.com/spreadsheets/d/17p_8fyyK8fT_UuAU78RRV16N_JrTxfmt-HCEKXivk9E/edit#gid=0) with some data from a new project that simply logs the fuzzer generated values, i.e. dict inputs are minimal so this is a worst case scenario. You can see with uints only 29% of values are unique, and 31% of values are the edge biased values. 

Given that only 29% of values are unique, this means 71k of the 100k runs were wasted execution.

For a fuzz test with `n` inputs, we should consider tracking each unique combination of inputs used to avoid re-using the same inputs. A few consideration:
- If we naively re-generate inputs when finding a duplicate, this might not be any better than just increasing the number of runs (performance-wise)
- A more performant approach may be to modify the strategies. I think each input of a fuzz test is generated independently, so we don't know if we have duplicate inputs until all values are chosen. We may be able to modify the strategies such that all required inputs are generated in the same strategy

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation

Closes #3521 

(this is a WIP, I see a substantial code refactoring coming in this area, atm just looking for some feedback re proposed implementation / UX. @mattsse @DaniPopes @klkvr @mds1 thank you)

As mentioned in #3521 the number of fuzzed edge cases is too high (only 29% unique), this is due to:
- int and uint strategies fall back to generating edge cases in case there's no fixture to use, see
https://github.com/foundry-rs/foundry/blob/master/crates/evm/fuzz/src/strategies/uint.rs#L114-L118
https://github.com/foundry-rs/foundry/blob/master/crates/evm/fuzz/src/strategies/int.rs#L130-L134
- defining fixtures is not supported so this gives a much higher weight for generating edge cases, see
https://github.com/foundry-rs/foundry/blob/master/crates/evm/fuzz/src/strategies/param.rs#L41
https://github.com/foundry-rs/foundry/blob/master/crates/evm/fuzz/src/strategies/param.rs#L46

## Solution

Proposed solution:
- fall back to random values in case there's no fixtures defined
- introduce concept and ability to programmatically define fixtures for named parameters in Solidity code. In first phase this applies to uint, int and address parameter types (for address replacing solution provided for #5868 in PR #7240) 

Fixtures can be defined as array of values to be applied in fuzzed tests, PR supports address, uint and int types but can be easily extended to other types:
### Fixtures sample:
- To define fixtures for `owner` fuzzed parameter of type `address`
```Solidity
address[] ownerFixture;

function setUp() public {
    ownerFixture.push(address(1));
    ownerFixture.push(address(2));
    ownerFixture.push(address(3));
}

function fixtures_owner() public returns (address[] memory) {
    return ownerFixture;
}
```
- write fuzzed tests by using `owner` param name so all defined fixtures are used when fuzzing
 ```Solidity
function testFuzz_Address(address owner) public {
}
```
- if a function returning different fixture type is declared, e.g.
 ```Solidity
function fixtures_owner() public returns (uint[] memory) {
}
```
then an error is displayed and fuzzing is performed with random generated addresses
- same definitions and rules applies for `uint` and `int` fixtures types

Some numbers collected from running without and with fixtures for a uint fuzzed param:

Run (with fixtures) | edge | random | fixtures 
--- | --- | --- | --- 
1 | 11.4% | 45.2% | 43.24% 
2 | 12.5% | 51.19% | 36.3% 
3 | 9.86% | 52.63% | 37.5%
4 | 12.92% | 43.53% | 43.53% 

Run (no fixtures) | edge | random 
--- | --- | ---
1 | 13.7% | 86.3% 
2 | 9.5% | 90.5%  
3 | 8.2% | 91.8% 
4 | 8.7% | 91.3%

### CODE REVIEW

File: crates/evm/fuzz/src/strategies/address.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+/// If fixture is not a valid type then an error is raised and test suite will continue to execute
+/// with random address.
+///
+///
+/// For example:
+/// To define fixture for `owner` fuzzed parameter, return an array of possible values from
+/// `function fixture_owner() public returns (address[] memory)`.
+/// Use `owner` named parameter in fuzzed test in order to create a custom strategy
+/// `function testFuzz_ownerAddress(address owner, uint amount)`.
+#[derive(Debug)]
```
Comment (mattsse) [inline crates/evm/fuzz/src/strategies/address.rs:21]:
```suggestion
#[derive(Debug, Default)]
#[non_exhaustive]
```

File: crates/forge/bin/cmd/test/mod.rs [hunk 2]
```diff
@@ -221,7 +221,7 @@ impl TestArgs {
             *test_pattern = Some(debug_test_pattern.clone());
         }
 
-        let outcome = self.run_tests(runner, config, verbosity, &filter).await?;
+        let outcome = self.run_tests(runner, config.clone(), verbosity, &filter).await?;
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/test/mod.rs:224]: These clones are not necessary

File: crates/forge/bin/cmd/test/mod.rs [hunk 3]
```diff
@@ -181,7 +181,7 @@ impl TestArgs {
 
         let test_options: TestOptions = TestOptionsBuilder::default()
             .fuzz(config.clone().fuzz)
-            .invariant(config.invariant)
+            .invariant(config.clone().invariant)
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/test/mod.rs:184]: move clone after .invariant to clone only that object instead of the entire config

same above for fuzz

File: crates/forge/src/runner.rs [hunk 4]
```diff
@@ -174,14 +174,54 @@ impl<'a> ContractRunner<'a> {
...
+    /// and followed by the name of the parameter.
+    ///
+    /// For example:
+    /// `fixture_test() returns (address[] memory)` function
+    /// define an array of addresses to be used for fuzzed `test` named parameter in scope of the
+    /// current test.
+    fn fuzz_fixtures(&mut self, address: Address) -> FuzzFixtures {
+        // collect test fixtures param:array of values
+        let mut fixtures = HashMap::new();
+        self.contract.functions().for_each(|func| {
```
Comment (DaniPopes) [inline crates/forge/src/runner.rs:210]: use a for loop with `.filter(|f| f.name.is_fixture())`

File: crates/forge/src/runner.rs [hunk 5]
```diff
@@ -174,14 +174,54 @@ impl<'a> ContractRunner<'a> {
...
+            if func.name.is_fixture() {
+                if let Ok(CallResult { raw: _, decoded_result }) =
+                    self.executor.call(CALLER, address, func, &[], U256::ZERO, None)
+                {
+                    fixtures.insert(
+                        func.name.strip_prefix("fixture_").unwrap().to_string(),
```
Comment (DaniPopes) [inline crates/forge/src/runner.rs:216]: we should check if there's a function with the stripped name and issue a warning

File: crates/forge/src/runner.rs [hunk 6]
```diff
@@ -174,14 +174,54 @@ impl<'a> ContractRunner<'a> {
...
+                        decoded_result,
```
Comment (DaniPopes) [inline crates/forge/src/runner.rs:217]: we should check for validity of the decoded result here instead of later when running the tests since all necessary info to do this is available here

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/forge/src/runner.rs (hunk 6 above)
Comment (grandizzy) [reply crates/forge/src/runner.rs]: that would be nice to have but needs special handling of handler based testing (where fuzzed functions are not in same contract that defines `setUp/invariant_/fixture_` functions). It can be done but it adds code complexity to look up such info in all contracts created when `setUp`, wdyt?

**Comment by grandizzy** [comment]
I addressed comments (but https://github.com/foundry-rs/foundry/pull/7428#discussion_r1533278017) and additionally moved `fuzz_calldata_from_state` from `state` mod in same `calldata` mod as `fuzz_calldata`, think it makes more sense to have grouped them like this. Please recheck

**Comment by grandizzy** [comment]
hey @mattsse @DaniPopes I added fixtures for string and bytes too and a Solidity test to cover all combinations, pls have a look when you get time. thank you

**Comment by mds1** [comment]
Fixture idea is neat! A few questions:
1. Is that only for fuzz tests, or also for invariant tests? If the latter, is it reevaluated after each call in a sequence?
2. Thoughts on trimming leading/trailing underscores, so if my argument name is `address _owner` or `address owner_` I don't need a funky looking fixture name of `fixture__owner` or `fixture_owner_`? With this approach we'd also want to error if argument names are duplicated after trimming
3. Variable names are now significant as they must be unique when you want to restrict the set of tests a fixture applies to. This feels a bit unintuitive with no precedent in foundry. Perhaps instead of the current approach, the mapping for arg name to fixture function name should be specified via inline-config comments (since it's supported at the test and contract level) to allow for this. This also remedies (2) without needing to trim names. Open to thoughts/ideas here. 
4. We should error if a fixture matches no known args (or if we go with (3), if a fixture is never used)
5. Can you provide a concrete example of when to use this, e.g. how it helps with #5868

**Comment by grandizzy** [comment]
Thanks for comments! Pls see inline

> Fixture idea is neat! A few questions:
> 
>     1. Is that only for fuzz tests, or also for invariant tests? If the latter, is it reevaluated after each call in a sequence?
> 

That's for both fuzz and invariant tests and applies only for param fuzzing (doesn't apply to fuzzing from state). The strategy is an index strategy which randomly picks an item from fixtures so basically reevaluated.

>     2. Thoughts on trimming leading/trailing underscores, so if my argument name is `address _owner` or `address owner_` I don't need a funky looking fixture name of `fixture__owner` or `fixture_owner_`? With this approach we'd also want to error if argument names are duplicated after trimming
> 

Ah, yeah, good point, missed the case where params could have `_` in their name. I like proposed solution in point 3 (inline config) to avoid this.

>     3. Variable names are now significant as they must be unique when you want to restrict the set of tests a fixture applies to. This feels a bit unintuitive with no precedent in foundry. Perhaps instead of the current approach, the mapping for arg name to fixture function name should be specified via inline-config comments (since it's supported at the test and contract level) to allow for this. This also remedies (2) without needing to trim names. Open to thoughts/ideas here.
> 

Agree, will implement such.

>     4. We should error if a fixture matches no known args (or if we go with (3), if a fixture is never used)

Can do that but will require pre processing and checking all functions/params from all handlers which could be a little bit ugly. Will check how the code looks like with adding such.

> 
>     5. Can you provide a concrete example of when to use this, e.g. how it helps with [Invariant violation not found on simple 2 step owner transfer test #5868](https://github.com/foundry-rs/foundry/issues/5868)

This approach helps when you want to be sure that certain values are used when fuzzing functions, for example in #5868 you want to make sure same set of addresses are included in fuzzed values, so one could just reuse same fixture
https://github.com/foundry-rs/foundry/blob/14273fac59fcaf3354b82b0df6bb39b1208c3985/testdata/default/fuzz/invariant/common/InvariantCalldataDictionary.t.sol#L84-L90
then `acceptOwnership` will be called at some point with `candidate` address as it is using the same fixture
https://github.com/foundry-rs/foundry/blob/14273fac59fcaf3354b82b0df6bb39b1208c3985/testdata/default/fuzz/invariant/common/InvariantCalldataDictionary.t.sol#L48-L59
Basically is to make actor management in invariant testing easier as right now one should use approach explained in https://book.getfoundry.sh/forge/invariant-testing#actor-management. (while with fixtures this won't be needed anymore).
Another example would be if you want let's say to be sure certain numbers from Fibonacci sequence are used in fuzzing (see also this test sample added with this PR https://github.com/foundry-rs/foundry/blob/14273fac59fcaf3354b82b0df6bb39b1208c3985/testdata/default/fuzz/invariant/common/InvariantFixtures.t.sol)
Or could even help with #2851 maybe?

**Comment by grandizzy** [comment]
@mds1 I implemented suggested inline conf, looks much better this way. Re `4. We should error if a fixture matches no known args (or if we go with (3), if a fixture is never used)` that's a little bit hard to do at the time we collect fixtures as there could be multiple handlers and defined in different files than test setup, so it needs to wait somehow until test setup. Since this one gets a little bit beefy, I suggest to follow up improvement in different PR if you OK with, lmk and will move this back to ready to review

**Comment by mds1** [comment]
Thanks a lot for all the answers and changes :) 

Totally ok with doing the last one in a follow up and marking this as ready for review. We might be able to do something similar to what we do for `--match-test`, where the first pass when compiling would only reqest natspec+ABI output for the checks, and a second pass requests the bytecode and other outputs. @klkvr might have thoughts here too

File: crates/forge/bin/cmd/test/mod.rs [hunk 7]
```diff
@@ -202,7 +202,7 @@ impl TestArgs {
         let runner = MultiContractRunnerBuilder::default()
             .set_debug(should_debug)
             .initial_balance(evm_opts.initial_balance)
-            .evm_spec(config.evm_spec_id())
+            .evm_spec(config.clone().evm_spec_id())
```
[the remark this answers is not in the document]
Comment (grandizzy) [reply crates/forge/bin/cmd/test/mod.rs]: sorry, missed this comment, removed with https://github.com/foundry-rs/foundry/pull/7428/commits/a5e8da0935a112b380ab330ec878c7d44ca16dd0

File: crates/forge/src/runner.rs [hunk 8]
```diff
@@ -165,14 +165,53 @@ impl<'a> ContractRunner<'a> {
...
     }
 
+    /// Collect fixtures from test contract. Fixtures are functions prefixed with `fixture_` key
+    /// and followed by the name of the parameter.
+    ///
+    /// For example:
+    /// `fixture_test() returns (address[] memory)` function
+    /// define an array of addresses to be used for fuzzed `test` named parameter in scope of the
+    /// current test.
+    fn fuzz_fixtures(&mut self, address: Address, fixtures: &InlineFixturesConfig) -> FuzzFixtures {
```
Comment (klkvr) [inline crates/forge/src/runner.rs:198]: Shouldn't this be more strict considering that we no more require `fixture_` prefix?

File: crates/evm/fuzz/src/strategies/string.rs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+impl StringStrategy {
+    /// Create a new string strategy.
+    pub fn init(fixtures: Option<&[DynSolValue]>) -> BoxedStrategy<DynSolValue> {
+        let value_from_fixture = |fixture: Option<&DynSolValue>| {
+            if let Some(fixture) = fixture {
+                if let Some(fixture) = fixture.as_str() {
+                    return DynSolValue::String(fixture.to_string());
+                }
+            }
```
Comment (klkvr) [inline crates/evm/fuzz/src/strategies/string.rs:33]:
```suggestion
            if let Some(val @ DynSolValue::String(_)) = fixture {
                return val.clone()
            }
```

I believe the same is possible for address and bytes strats. Wondering why do we have different logic for value-generation when no fixtures are added vs when fixture returns incorrect value?

File: crates/evm/fuzz/src/lib.rs [hunk 10]
```diff
@@ -272,3 +272,28 @@ impl FuzzedCases {
...
+
+impl FuzzFixtures {
+    pub fn new(fixtures: HashMap<String, DynSolValue>) -> FuzzFixtures {
+        Self { inner: Arc::new(fixtures) }
+    }
+
+    /// Returns configured fixtures for `param_name` fuzzed parameter.
+    pub fn param_fixtures(&self, param_name: &String) -> Option<&[DynSolValue]> {
+        if let Some(param_fixtures) = self.inner.get(param_name) {
+            param_fixtures.as_array()
```
Comment (klkvr) [inline crates/evm/fuzz/src/lib.rs:294]: This will now only parse dynamic arrays, let's support fixed-sized arrays as well

**Review by klkvr — COMMENTED** [review]
sorry, just got a chance to read through, looking great! a couple questions

unfortunately Solidity does not offer great UX for dealing with arrays, so I think we should also consider to add support for reading fixtures from storage arrays:
```solidity
contract Test {
    uint256[] public fixureName = [1, 2, 3];
}
```

vs

```solidity
contract Test {
    function fixureName() public view returns(uint256[] memory vals) {
        vals = new uint256[](3);
        vals[0] = 1;
        vals[1] = 2;
        vals[2] = 3;
    }
}
```

File: crates/evm/fuzz/src/lib.rs (hunk 10 above)
Comment (grandizzy) [reply crates/evm/fuzz/src/lib.rs]: added with https://github.com/foundry-rs/foundry/pull/7428/commits/ca4d44b67d0a950670f09e94766a34750e1f537a

**Comment by grandizzy** [comment]
> > sorry, just got a chance to read through, looking great! a couple questions
> > unfortunately Solidity does not offer great UX for dealing with arrays, so I think we should also consider adding support for reading fixtures from storage arrays:
> > ```solidity
> > contract Test {
> >     uint256[] public fixureName = [1, 2, 3];
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > vs
> > ```solidity
> > contract Test {
> >     function fixureName() public view returns(uint256[] memory vals) {
> >         vals = new uint256[](3);
> >         vals[0] = 1;
> >         vals[1] = 2;
> >         vals[2] = 3;
> >     }
> > }
> > ```
> 
> yep, good call, I think this way we could even drop the inline config and do the best effort to match / map fixtures from functions within test contract, will look into adding such

@klkvr I added support for reading fixtures from storage and dropped the inline fixtures config with https://github.com/foundry-rs/foundry/pull/7428/commits/fb86084fe7e283b62deb0beeed36ac0a05d329fc Didn't find a nicer way to populate fixtures from storage array but to call function with incremented indexes, pls lmk if you see a nicer way for doing this. see
https://github.com/foundry-rs/foundry/pull/7428/commits/fb86084fe7e283b62deb0beeed36ac0a05d329fc#diff-98c09a5d1cb595f37109d7b1c0109ee32dd74dc1d253cdda9a63e4ce32bf49d5R227-R251

File: crates/forge/src/runner.rs (hunk 8 above)
Comment (grandizzy) [reply crates/forge/src/runner.rs]: yeah, updated comment to reflect also the possibility of fixtures from storage arrays

File: crates/evm/fuzz/src/strategies/string.rs (hunk 9 above)
Comment (grandizzy) [reply crates/evm/fuzz/src/strategies/string.rs]: thing is that we don't know if fixture is of valid type until the point we generate value from strategy, so cannot use default `DynSolValue::type_strategy` in this case. Wonder if it would be better to not generate random values and just let fuzzer panic if wrong type (so user will be aware right away that configured fixtures are not used in campaign) instead printing errors, wdyt?

Comment (klkvr) [reply crates/evm/fuzz/src/strategies/string.rs]: I think we should change flow a bit so that `ContractRunner::fuzz_fixtures` flattens all arrays right away and keeps sets of `DynSolValue` for each func name, enforcing that all of them have same type.

That way, when creating strategy we can validate that all matching fixtures for given parameter have the same value as the parameter itself. That way we can just panic if we got a fixture of invalid type and this should be safe

Comment (grandizzy) [reply crates/evm/fuzz/src/strategies/string.rs]: yeah, agree, I just changed code to let fuzzer panic if the param type is not of fixtures type. The risk here was when fixtures were matched by param name only (as in someone could have been declaring a `string[]` fixture for fuzzing an `address` type) but now since we explicitly require functions prefixed with `fixture` I don't think that's a problem anymore - this also simplified code a lot.

**Review by klkvr — CHANGES_REQUESTED** [review]
@grandizzy we shouldn't just panic on user errors, I meant that we should also have validation somewhere which will throw a error, thus making panic more safe

we can probably either do another pass through ABIs before running tests to check if there are any mismatches between fixture types and parameter types or validate this when generating strategies.

File: crates/forge/src/runner.rs [hunk 11]
```diff
@@ -165,14 +166,84 @@ impl<'a> ContractRunner<'a> {
...
+    /// Storage array fixtures:
+    /// `uint256[] public fixture_amount = [1, 2, 3];`
+    /// define an array of uint256 values to be used for fuzzing `amount` named parameter in scope
+    /// of the current test.
+    ///
+    /// Function fixtures:
+    /// `function fixture_owner() public returns (address[] memory){}`
+    /// returns an array of addresses to be used for fuzzing `owner` named parameter in scope of the
+    /// current test.
+    fn fuzz_fixtures(&mut self, address: Address) -> FuzzFixtures {
```
Comment (klkvr) [inline crates/forge/src/runner.rs:208]: I think we can just have this check here

**Comment by grandizzy** [comment]
> @grandizzy we shouldn't just panic on user errors, I meant that we should also have validation somewhere which will throw a error, thus making panic more safe
> 
> we can probably either do another pass through ABIs before running tests to check if there are any mismatches between fixture types and parameter types or validate this when generating strategies.

@klkvr if I am not missing something a fully validation when collecting fixtures from invariant contract isn't possible as there could be targets created during runs that we don't have at the moment test is set up. Even if we manage to do the validation then the option would be to panic if params not of same types (otherwise it will still panic later in the run). If my rationale is correct what do you think if we just use some default values in case fixture of different type (like 0 for uint, adrress(0), empty string, etc.) and raise errors instead panic, they are quite noisy in the console and hard to be missed, so dev will know fixtures needs to be amended. thanks!

**Comment by klkvr** [comment]
> @klkvr if I am not missing something a fully validation when collecting fixtures from invariant contract isn't possible as there could be targets created during runs that we don't have at the moment test is set up

hmm, I see. what if we do `prop_filter_map` and reject such values thus forcing fallback to default strategy?

**Comment by grandizzy** [comment]
> > @klkvr if I am not missing something a fully validation when collecting fixtures from invariant contract isn't possible as there could be targets created during runs that we don't have at the moment test is set up
> 
> hmm, I see. what if we do `prop_filter_map` and reject such values thus forcing fallback to default strategy? also display a error/warning to user

oh, awesome stuff, this should handle all cases, added in https://github.com/foundry-rs/foundry/pull/7428/commits/997c5d4ef994aa11ddaa4e0be90632e36ea2d3dc

File: crates/evm/fuzz/src/strategies/mod.rs [hunk 12]
```diff
@@ -8,14 +8,38 @@ mod param;
...
+                        fixtures.iter().enumerate().map(|(_, value)| value.to_owned()).collect();
+                    let custom_fixtures_len = custom_fixtures.len();
+                    any::<prop::sample::Index>()
+                        .prop_filter_map("invalid fixture", move |index| {
+                            let index = index.index(custom_fixtures_len);
+                            $strategy_value(custom_fixtures.get(index))
+                    })
+                },
+                50 => $default_strategy
+            ].boxed()
```
Comment (klkvr) [inline crates/evm/fuzz/src/strategies/mod.rs:38]: shouldn't we apply `prop_filter_map` to the entier `prop_oneof!` output? I believe right now if we only have invalid fixtures it will fail with max rejects while it'd be better if it used default strategy instead

Comment (grandizzy) [reply crates/evm/fuzz/src/strategies/mod.rs:38]: was thinking that could be better to fail (faster) with max rejects rather than run tests with a config that is not really desired, but can change to apply to both strategies, wdyt?

Comment (klkvr) [reply crates/evm/fuzz/src/strategies/mod.rs:38]: right, let's keep it like this for now

**Review by klkvr — APPROVED** [review]
LGTM! nice work

@mattsse @DaniPopes mind taking a look

**Review by DaniPopes — APPROVED** [review]
